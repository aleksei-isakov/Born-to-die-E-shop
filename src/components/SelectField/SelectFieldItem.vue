<template>
  <li @mousedown="onClickChooseItem">
    {{ category }}
  </li>
</template>

<script>
export default {
  name: 'SelectFieldItem',

  props: {
    category: {
      type: String,
      required: true,
      default: ''
    },

    index: {
      type: Number,
      required: true,
      default: 0
    }
  },

  methods: {
    onClickChooseItem() {
      this.$emit('chooseItem', this.index);
    }
  }
};
</script>
