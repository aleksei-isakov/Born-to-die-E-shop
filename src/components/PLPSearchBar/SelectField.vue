<template>
  <div>
    <button
      class="select-field"
      @click="onClickToggleDropdown"
      @blur="onBlurCloseDropdown"
    >
      <div class="select-field__category">Category</div>
      <div>{{ currentCategory }}</div>
      <img
        class="select-field__arrow-icon"
        :src="arrowIcon"
        alt="arrow-icon"
        :class="{ reverse: isDropdownOpen }"
      />
    </button>
    <ul class="dropdown" :class="{ open: isDropdownOpen }">
      <select-field-item
        v-for="(category, i) in categories"
        :key="i"
        :category="category"
        :index="i"
        @chooseItem="onChooseItem"
      />
    </ul>
  </div>
</template>

<script>
import arrowIcon from '@/assets/Icons/arrow.svg';
import SelectFieldItem from './SelectFieldItem.vue';
import { mapActions } from 'vuex';

export default {
  name: 'SelectField',

  components: {
    SelectFieldItem
  },

  props: {
    categories: {
      type: Array,
      required: true,
      default: () => []
    }
  },

  data() {
    return {
      arrowIcon: arrowIcon,
      isDropdownOpen: false,
      activeCategoryIndex: 0
    };
  },

  computed: {
    currentCategory() {
      return this.categories[this.activeCategoryIndex];
    }
  },

  methods: {
    ...mapActions('PlpPageModule', ['setCurrentCategory']),

    onClickToggleDropdown() {
      this.isDropdownOpen = !this.isDropdownOpen;
    },

    onBlurCloseDropdown() {
      this.isDropdownOpen = false;
    },

    onChooseItem(index) {
      this.activeCategoryIndex = index;
      this.setCurrentCategory(this.currentCategory);
    }
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';

.select-field {
  margin-right: 20px;
  width: 15vw;
  min-width: 120px;
  height: 7vh;
  background-color: $white;
  border: 1px solid rgb(211, 210, 210);
  border-radius: 5px;
  position: relative;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 0 10px;
  cursor: pointer;
}

.select-field__category {
  background-color: $white;
  color: #b3b4b6;
  padding: 0 5px 0 5px;
  position: absolute;
  top: -8px;
  left: 10px;
  font-size: 12px;
}

.dropdown {
  background-color: $white;
  border: 1px solid rgb(211, 210, 210);
  border-radius: 5px;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  padding: 0px;
  width: 15vw;
  margin: 5px 0 0 0;
  opacity: 0;
  transition: $transition;
  height: 0px;
  min-width: 120px;
  max-height: 300px;
  overflow: auto;
  position: relative;
  z-index: 2;
}

@media screen and (max-width: $mobile-size) {
  .dropdown {
    width: 90vw;
    position: absolute;
  }
  .select-field {
    width: 90vw;
    margin: 0;
  }
}

.dropdown li {
  padding: 10px 10px;
  list-style: none;
  width: 100%;
  text-align: left;
  cursor: pointer;
}

.dropdown li:hover {
  background-color: rgb(238, 237, 237);
}

.select-field__arrow-icon {
  width: 10px;
  justify-self: right;
}

.open {
  height: fit-content;
  opacity: 1;
}

.reverse {
  transform: rotateX(180deg);
}
</style>
