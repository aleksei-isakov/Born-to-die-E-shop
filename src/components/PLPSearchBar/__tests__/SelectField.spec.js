import SelectField from '../SelectField.vue';
import { mount } from '@vue/test-utils';

let wrapper;

beforeEach(() => {
  wrapper = mount(SelectField, {
    propsData: {
      categories: ['a', 'b', 'c']
    },

    data() {
      return {
        isDropdownOpen: false
      };
    }
  });
});

afterEach(() => {
  wrapper.destroy();
});

describe('SelectField', () => {
  it('should change visibility of dropdown by click', () => {
    const button = wrapper.find('.select-field');

    button.trigger('click');
    expect(wrapper.vm.isDropdownOpen).toBe(true);
    button.trigger('click');
    expect(wrapper.vm.isDropdownOpen).toBe(false);
  });

  it('should render correctly and match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
