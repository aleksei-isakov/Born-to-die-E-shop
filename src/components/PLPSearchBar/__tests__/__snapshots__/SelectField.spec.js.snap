// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectField should render correctly and match snapshot 1`] = `
<div>
  <button
    class="select-field"
  >
    <div
      class="select-field__category"
    >
      Category
    </div>
     
    <div>
      a
    </div>
     
    <img
      alt="arrow-icon"
      class="select-field__arrow-icon"
      src=""
    />
  </button>
   
  <ul
    class="dropdown"
  >
    <li>
      
  a

    </li>
    <li>
      
  b

    </li>
    <li>
      
  c

    </li>
  </ul>
</div>
`;
